### PR TITLE
chore: more advanced types for sync/async resource

### PR DESCRIPTION
**Pull request by dyladan**
This allows the resource on the span to be sync in the export pipeline while allowing it to be async in the tracer and removes the need for casting.

I only did tracing for the example, but the same should be possible for metrics.

### CODE REVIEW

File: packages/opentelemetry-tracing/src/Span.ts [hunk 1]
```diff
@@ -194,6 +190,13 @@ export class Span implements api.Span, ReadableSpan {
     return this._ended;
   }
 
+  get resource(): Resource {
```
Comment (dyladan) [inline packages/opentelemetry-tracing/src/Span.ts:226]: If the promise is resolved (after _resolveResources completes), then this will act like a regular sync resource to satisfy the ReadableSpan interface.

File: packages/opentelemetry-tracing/src/Span.ts [hunk 2]
```diff
@@ -23,15 +23,14 @@ import {
...
-import { ReadableSpan } from './export/ReadableSpan';
-import { Tracer } from './Tracer';
 import { SpanProcessor } from './SpanProcessor';
+import { Tracer } from './Tracer';
 import { TraceParams } from './types';
 
 /**
  * This class represents a span.
  */
-export class Span implements api.Span, ReadableSpan {
```
Comment (obecny) [inline packages/opentelemetry-tracing/src/Span.ts:34]: why did you remove ReadableSpan ?

File: packages/opentelemetry-tracing/src/Span.ts (hunk 1 above)
Comment (obecny) [reply packages/opentelemetry-tracing/src/Span.ts:226]: Imagine the following situation
1. Resource is provided as promise
2. Start span
3. Something is checking for resource, then just started span receives a resource which is `Resource.createTelemetrySDKResource()`
4. Resource that was provided as a promise has been just resolved
5. Span ended - export span
6. Span is being exported with wrong resource instead of the one provided from promise

Another situation
1. Resource is provided as promise
2. Start span
3. Span ended - export span
4. Span is being exported with wrong resource instead of the one provided from promise
5. Resource that was provided as a promise has been just resolved - but previously created span has been already exported with wrong resource

Comment (dyladan) [reply packages/opentelemetry-tracing/src/Span.ts:226]: >Something is checking for resource

Checking for resource isn't even a part of the public API until we get a "Readable Span", which is not the same as a span. It might be easier if we just send a plain object instead of the actual Span with a getter?

>Span is being exported with wrong resource instead of the one provided from promise

No, the export will still have the correct resource as the span will not be sent to the exporter until the resource is resolved, and getters don't cache. Try it if you don't believe me.

>Another situation

>Span is being exported with wrong resource instead of the one provided from promise

Why? The span isn't actually exported until the resource resolves.

File: packages/opentelemetry-tracing/src/Span.ts (hunk 2 above)
Comment (dyladan) [reply packages/opentelemetry-tracing/src/Span.ts:34]: It can be added back.

File: packages/opentelemetry-tracing/src/Span.ts (hunk 1 above)
Comment (dyladan) [reply packages/opentelemetry-tracing/src/Span.ts:226]: ?

Comment (obecny) [reply packages/opentelemetry-tracing/src/Span.ts:226]: I wrote you a message on gitter 4 days ago, please read :)

Comment (dyladan) [reply packages/opentelemetry-tracing/src/Span.ts:226]: D: ok thanks

Too many communication channels to keep track of

**Comment by dyladan** [comment]
Closed for https://github.com/open-telemetry/opentelemetry-js/pull/1484
